src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Handler/Completion/CompletionHandler.cs

```diff
@@ -99,10 +109,29 @@ private static CompletionItem CreateCompletionItem(XamlCompletionItem xamlComple
                     Title = CreateEventHandlerCommandTitle
                 };
             }
+            else if (xamlCompletion.RetriggerCompletion)
+            {
+                // Retriger completion after commit
+                item.Command = new Command()
+                {
+                    CommandIdentifier = StringConstants.RetriggerCompletionCommand,
+                    Title = RetriggerCompletionCommandTitle
+                };
+            }
 
             return item;
         }
 
+        private static SumType<string[], CommitCharacter[]> GetCommitCharacters(XamlCompletionItem completionItem)
+        {
+            if(completionItem.XamlCommitCharacters == null)
+            {
+                return completionItem.CommitCharacters;
+            }
+
+            return completionItem.XamlCommitCharacters.Select(c => new CommitCharacter() { Character = c.Character.ToString(), Insert = c.Insert }).ToArray();
```

Comment:
you may want to cache these similar to what we do here to avoid allocations on every item
http://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Handler/Completion/CompletionHandler.cs,294

Reply:
That's a good feedback and thanks for bring it up. I went back and looked at this. We are not able to do exact as what Roslyn is doing as this new CommitCharacter requires information more than just the characters, but I get the idea of caching. As XAML generates commit characters base on completion item kind I think caching the characters by kind should be sufficient here.

Reply:
Could you make the commitCharactersCache a class member?

Reply:
I have thought about this, but I also see having the cache base on each completion session could probably give us more flexibilities as if later some completion sessions may need to have different set of commit characters we don't need to fix the roslyn code. Also I am trying to align to what Roslyn is doing. Roslyn has the cache for each session so I think it should be fine performance wise.